File: src/gui/widgets/inferencewidget.cpp
#include "inferencewidget.h"
#include "imageviewerwidget.h"
#include <QFileDialog>
#include <QListWidgetItem>
#include <QProgressBar>
#include <QGroupBox>
#include <QFormLayout>

const QStringList InferenceWidget::m_classNames = {
    "T-shirt/top", "Trouser", "Pullover", "Dress", "Coat",
    "Sandal", "Shirt", "Sneaker", "Bag", "Ankle boot"
};

InferenceWidget::InferenceWidget(QWidget *parent)
    : QWidget(parent)
    , m_imageViewer(nullptr)
    , m_predictedClassLabel(nullptr)
    , m_confidenceLabel(nullptr)
    , m_probabilityList(nullptr)
    , m_runInferenceButton(nullptr)
    , m_loadImageButton(nullptr)
    , m_numSamplesSpinBox(nullptr)
    , m_inferenceProgress(nullptr)
{
    setupUI();
}

void InferenceWidget::setupUI()
{
    QHBoxLayout* mainLayout = new QHBoxLayout(this);
    
    QGroupBox* imageGroup = new QGroupBox("Input Image", this);
    QVBoxLayout* imageLayout = new QVBoxLayout(imageGroup);
    
    m_imageViewer = new ImageViewerWidget(this);
    m_imageViewer->setMinimumSize(300, 300);
    imageLayout->addWidget(m_imageViewer);
    
    QHBoxLayout* buttonLayout = new QHBoxLayout();
    m_loadImageButton = new QPushButton("Load Image", this);
    connect(m_loadImageButton, &QPushButton::clicked, this, &InferenceWidget::onLoadImage);
    buttonLayout->addWidget(m_loadImageButton);
    buttonLayout->addStretch();
    imageLayout->addLayout(buttonLayout);
    
    mainLayout->addWidget(imageGroup, 1);
    
    QGroupBox* resultsGroup = new QGroupBox("Inference Results", this);
    QVBoxLayout* resultsLayout = new QVBoxLayout(resultsGroup);
    
    m_predictedClassLabel = new QLabel("Predicted: -", this);
    m_predictedClassLabel->setStyleSheet("QLabel { font-size: 18pt; font-weight: bold; padding: 10px; }");
    resultsLayout->addWidget(m_predictedClassLabel);
    
    m_confidenceLabel = new QLabel("Confidence: -", this);
    m_confidenceLabel->setStyleSheet("QLabel { font-size: 14pt; padding: 5px; }");
    resultsLayout->addWidget(m_confidenceLabel);
    
    resultsLayout->addWidget(new QLabel("Class Probabilities:", this));
    m_probabilityList = new QListWidget(this);
    m_probabilityList->setMaximumHeight(300);
    resultsLayout->addWidget(m_probabilityList);
    
    QGroupBox* inferenceGroup = new QGroupBox("Inference Settings", this);
    QFormLayout* inferenceLayout = new QFormLayout(inferenceGroup);
    
    m_numSamplesSpinBox = new QSpinBox(this);
    m_numSamplesSpinBox->setMinimum(1);
    m_numSamplesSpinBox->setMaximum(10000);
    m_numSamplesSpinBox->setValue(100);
    inferenceLayout->addRow("Number of Samples:", m_numSamplesSpinBox);
    
    m_runInferenceButton = new QPushButton("Run Inference", this);
    m_runInferenceButton->setStyleSheet("QPushButton { background-color: #2196F3; color: white; font-weight: bold; padding: 10px; }");
    connect(m_runInferenceButton, &QPushButton::clicked, this, &InferenceWidget::onRunInference);
    inferenceLayout->addRow(m_runInferenceButton);
    
    m_inferenceProgress = new QProgressBar(this);
    m_inferenceProgress->setVisible(false);
    inferenceLayout->addRow("Progress:", m_inferenceProgress);
    
    resultsLayout->addWidget(inferenceGroup);
    resultsLayout->addStretch();
    
    mainLayout->addWidget(resultsGroup, 1);
}

void InferenceWidget::displayPrediction(int predictedClass, const QImage& image, const QVector<double>& probabilities)
{
    if (m_imageViewer) {
        m_imageViewer->displayImage(image);
    }
    
    if (predictedClass >= 0 && predictedClass < m_classNames.size()) {
        QString className = m_classNames[predictedClass];
        m_predictedClassLabel->setText(QString("Predicted: %1").arg(className));
        
        if (probabilities.size() > predictedClass) {
            double confidence = probabilities[predictedClass];
            m_confidenceLabel->setText(QString("Confidence: %1%").arg(confidence * 100, 0, 'f', 2));
        }
    }
    
    updateClassProbabilities(probabilities);
}

void InferenceWidget::updateClassProbabilities(const QVector<double>& probabilities)
{
    m_probabilityList->clear();
    
    QVector<QPair<int, double>> sortedProbs;
    for (int i = 0; i < probabilities.size() && i < m_classNames.size(); ++i) {
        sortedProbs.append(qMakePair(i, probabilities[i]));
    }
    
    std::sort(sortedProbs.begin(), sortedProbs.end(), 
        [](const QPair<int, double>& a, const QPair<int, double>& b) {
            return a.second > b.second;
        });
    
    for (const auto& pair : sortedProbs) {
        int classIdx = pair.first;
        double prob = pair.second;
        QString itemText = QString("%1: %2%")
            .arg(m_classNames[classIdx])
            .arg(prob * 100, 0, 'f', 2);
        
        QListWidgetItem* item = new QListWidgetItem(itemText, m_probabilityList);
        item->setData(Qt::UserRole, classIdx);
        m_probabilityList->addItem(item);
    }
}

void InferenceWidget::onRunInference()
{
    emit onInferenceStarted();
    m_runInferenceButton->setEnabled(false);
    m_inferenceProgress->setVisible(true);
    m_inferenceProgress->setRange(0, 0);
}

void InferenceWidget::onLoadImage()
{
    QString filePath = QFileDialog::getOpenFileName(this, 
        "Load Fashion-MNIST Image", "", "Image Files (*.png *.jpg *.jpeg *.bmp);;All Files (*.*)");
    if (filePath.isEmpty()) return;
    
    QImage image(filePath);
    if (!image.isNull() && m_imageViewer) {
        m_imageViewer->displayImage(image);
    }
}

void InferenceWidget::onImageSelected(int index)
{
    // Handle image selection from dataset
}

void InferenceWidget::onInferenceStarted()
{
    m_runInferenceButton->setEnabled(false);
    m_inferenceProgress->setVisible(true);
}

void InferenceWidget::onInferenceFinished()
{
    m_runInferenceButton->setEnabled(true);
    m_inferenceProgress->setVisible(false);
}

File: src/gui_py/widgets/inferencewidget.py
from PyQt6.QtWidgets import (QWidget, QVBoxLayout, QHBoxLayout, QPushButton, 
                             QLabel, QGroupBox, QListWidget, QListWidgetItem,
                             QFormLayout, QSpinBox, QProgressBar, QFileDialog)
from PyQt6.QtCore import Qt, pyqtSignal
from PyQt6.QtGui import QImage
from .imageviewerwidget import ImageViewerWidget

class InferenceWidget(QWidget):
    onInferenceStarted = pyqtSignal()

    def __init__(self, controller):
        super().__init__()
        self.controller = controller
        self.dataset_type = "generic"
        self.current_image = None
        self.init_ui()
        
    def set_dataset_type(self, dataset_type):
        self.dataset_type = dataset_type
        
    def init_ui(self):
        main_layout = QHBoxLayout(self)
        
        # Left Side: Image
        image_group = QGroupBox("Input Image")
        image_layout = QVBoxLayout(image_group)
        
        self.image_viewer = ImageViewerWidget()
        self.image_viewer.setMinimumSize(300, 300)
        image_layout.addWidget(self.image_viewer)
        
        btn_layout = QHBoxLayout()
        self.load_btn = QPushButton("Load Image")
        self.load_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.load_btn.clicked.connect(self.on_load_image)
        btn_layout.addWidget(self.load_btn)
        btn_layout.addStretch()
        image_layout.addLayout(btn_layout)
        
        main_layout.addWidget(image_group, 1)
        
        # Right Side: Results
        results_group = QGroupBox("Inference Results")
        results_layout = QVBoxLayout(results_group)
        
        self.pred_label = QLabel("Predicted: -")
        self.pred_label.setStyleSheet("QLabel { font-size: 18pt; font-weight: bold; padding: 10px; }")
        results_layout.addWidget(self.pred_label)
        
        self.conf_label = QLabel("Confidence: -")
        self.conf_label.setStyleSheet("QLabel { font-size: 14pt; padding: 5px; }")
        results_layout.addWidget(self.conf_label)
        
        results_layout.addWidget(QLabel("Class Probabilities:"))
        self.prob_list = QListWidget()
        self.prob_list.setMaximumHeight(300)
        results_layout.addWidget(self.prob_list)
        
        # Inference Settings
        inf_group = QGroupBox("Inference Settings")
        inf_layout = QFormLayout(inf_group)
        
        self.num_samples = QSpinBox()
        self.num_samples.setRange(1, 10000)
        self.num_samples.setValue(100)
        inf_layout.addRow("Number of Samples:", self.num_samples)
        
        self.run_btn = QPushButton("Run Inference")
        self.run_btn.setStyleSheet("QPushButton { background-color: #2196F3; color: white; font-weight: bold; padding: 10px; }")
        self.run_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self.run_btn.clicked.connect(self.on_run_inference)
        inf_layout.addRow(self.run_btn)
        
        self.progress = QProgressBar()
        self.progress.setVisible(False)
        inf_layout.addRow("Progress:", self.progress)
        
        results_layout.addWidget(inf_group)
        results_layout.addStretch()
        
        main_layout.addWidget(results_group, 1)

    def on_load_image(self):
        # Start at src/data/samples if exists
        import os
        from pathlib import Path
        
        # We assume we are in src/gui_py/widgets/
        # Project root is ../../../
        current_dir = Path(__file__).parent
        project_root = current_dir.parent.parent.parent
        samples_dir = project_root / "src" / "data" / "samples"
        
        start_dir = str(samples_dir) if samples_dir.exists() else ""
        
        path, _ = QFileDialog.getOpenFileName(self, "Load Fashion-MNIST/MNIST Image", start_dir, 
                                            "Image Files (*.png *.jpg *.jpeg *.bmp);;All Files (*.*)")
        if path:
            image = QImage(path)
            if not image.isNull():
                self.current_image = image
                self.image_viewer.displayImage(image)

    def on_run_inference(self):
        if not self.current_image:
           return
           
        self.onInferenceStarted.emit()
        self.run_btn.setEnabled(False)
        self.progress.setVisible(True)
        self.progress.setRange(0, 0) # Indeterminate
        
        # Request controller to run inference with image
        self.controller.requestInference.emit(self.current_image)

    def displayPrediction(self, predicted_class, image, probabilities):
        if not image.isNull():
            self.image_viewer.displayImage(image)
            
        from dataset_utils import get_class_name
        
        # We default to generic or update if we know dataset
        name = get_class_name(self.dataset_type, predicted_class)
        self.pred_label.setText(f"Predicted: {name}")
            
        if probabilities and predicted_class < len(probabilities):
            conf = probabilities[predicted_class]
            self.conf_label.setText(f"Confidence: {conf*100:.2f}%")
                
        self.update_probabilities(probabilities)
        self.inference_finished() # Hide progress

    def update_probabilities(self, probabilities):
        self.prob_list.clear()
        if not probabilities: return
        
        from dataset_utils import get_class_name
        
        # Pair index with prob
        indexed_probs = [(i, p) for i, p in enumerate(probabilities)]
        # Sort desc by prob
        indexed_probs.sort(key=lambda x: x[1], reverse=True)
        
        for idx, prob in indexed_probs:
            name = get_class_name(self.dataset_type, idx) # Improve detection later
            text = f"{name}: {prob*100:.2f}%"
            item = QListWidgetItem(text)
            self.prob_list.addItem(item)

    def inference_finished(self):
        self.run_btn.setEnabled(True)
        self.progress.setVisible(False)
